### DISCUSSION TITLE
16f877a bug

### DISCUSSION THREAD

**Discussion by era62** [lcgamboa/picsimlab#105] [General]
hi, i'm new of this forum (gianandrea from capri's isle italy) so sorry if this issue is known.  my code starts with a loop reading pins ra4 and ra5 on a 16f877a.  this pin are pulled up and the buttons are "active low". on the real board and the proteus emulation it works fine. using picsimlab (breadboard) on windows or on linux, the code goes on as ra5 is low. it's very strange that selecting microcontroller pic 887 instead of 877a my code works fine on this emulator. lot of thanks for your software. best rgs.

**Comment by lcgamboa** [comment]
Hi @era62 ,

I have tested PIC16F877A with the code below and it worked without error.
```
#pragma config FOSC = HS        // Oscillator Selection bits (HS oscillator)
#pragma config WDTE = OFF       // Watchdog Timer Enable bit (WDT disabled)
#pragma config PWRTE = OFF      // Power-up Timer Enable bit (PWRT disabled)
#pragma config BOREN = OFF      // Brown-out Reset Enable bit (BOR disabled)
#pragma config LVP = OFF        // Low-Voltage (Single-Supply) In-Circuit Serial Programming Enable bit (RB3 is digital I/O, HV on MCLR must be used for programming)
#pragma config CPD = OFF        // Data EEPROM Memory Code Protection bit (Data EEPROM code protection off)
#pragma config WRT = OFF        // Flash Program Memory Write Enable bits (Write protection off; all program memory may be written to by EECON control)
#pragma config CP = OFF         // Flash Program Memory Code Protection bit (Code protection off)

#include <xc.h>

void main() {
    ADCON1 = 0x06;
    TRISAbits.TRISA4 = 1;
    TRISAbits.TRISA5 = 1;
    TRISB = 0;
    while (1) {

        if (PORTAbits.RA4 == 0) {
            PORTB = 0xF0;
        }
        if (PORTAbits.RA5 == 0) {
            PORTB = 0x0F;
        }
    }
}
```
From the description of your problem, you probably did not configure RA5 as a digital input (`ADCON1 = 0x06;`). PICSimLab is a digital simulator with limited support for analog signals, so for an analog pin an analog stimulus such as a potentiometer is required. To read the buttons, the pin must be configured as digital. The PinViewer tool from the Tools menu usually helps identify these types of problems.

**Reply by era62** [reply]
thanks for your answer. it may be so. maybe the proteus (as the real board) can read the 1 state on an analogic input. but i'm still confused because  i've used " declare all_digital = true" that normally worked on proton basic. may be a bug of proton. i've also on the datasheet seen that pic 16f887 has the same double use of port a. and my code works on your emulator setting pic 16f887. if you have time my code is on www.casamonetacapri.com/kaletombola3.asm. have a nice weekend.ga

**Comment by lcgamboa** [comment]
I could not assemble your code with mpasm, could you send the .hex file? 
Looking at the code, `declare all_digital = true` seems to configure the pins as digital correctly.

**Reply by era62** [reply]
i've found the mistake i made. i've put a remark on the line where i've set the ports a as input with tris command. the strange thing is that  on the real board and  on proteus my program was working the same reading the buttons. your emu is more perfect than the controller itself :-).i'm jocking.maybe the controller reads the port by default and the emu not. all my best,ga

**Reply by lcgamboa** [reply]
The correct thing is for the pins to be configured as input after the reset. I'll check in the simulator if this is being done. 
Thanks for sharing the problem.

**Reply by era62** [reply]
thank you for the answers and, of course for your great software, please let me know if the problem was the i/o default. regards

**Comment by lcgamboa** [comment]
Hi @era62 ,

I checked and by default the pins are set. I was unable to reproduce the issue. Please send the .hex file of the program that presents the problem.

**Reply by era62** [reply]
hi, the issue is the missing menu to choose the game.
www.casamonetacapri.com/untitled.pzw 
www.casamonetacapri.com/kaletombola3.hex has the issue. starts directly with the tombola game without let you choose the game. works fine if you choose pic 16f887 instead.
www.casamonetacapri.com/kaletombola4.hex is the working version (porta set as input)
thank you.

**Reply by lcgamboa** [reply]
Thank you for the code example. I have managed to fix the bug.  It was a problem with the TRISx mirror. This was fixed in PICSim commit [cf05ed7](https://github.com/lcgamboa/picsim/commit/cf05ed7becfe7df50b86947a5cb4a43f0efa3b16).

**Reply by era62** [reply]
thank you. have a nice day.
